include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,767 @@
+ *        while leaving the underlying sequence unmodified.
+ * \tparam inner_type The type of range that will be decorated with gaps; must model std::ranges::View,
+ *                    std::ranges::RandomAccessRange and std::ranges::SizedRange.
+ * \implements seqan3::aligned_sequence_concept
+ * \ingroup decorator
+ *
+ * \details
+ *
+ * This class may be used whenever you want to store or compute an alignment. The underlying (ungapped) sequence
+ * remains unmodified, and is augmented with gap information. The seqan3::gap_decorator_anchor_set behaves just like a
+ * vector over a gapped alphabet when iterating over it, inserting/erasing gaps or accessing a position. The only
+ * difference lies in the performance and size overhead (see below).
+ *
+ *
+ * ### Performance
+ *
+ * **n** The length of the underlying sequence.
+ * **k** The number of contiguous gaps (not gap symbols).
+ * **l** The total number of gap symbols.
+ *
+ * |            | access next | random access    | gap insert/erase at end | gap insert/erase random |
+ * |------------|-------------|----------------- |-------------------------|-------------------------|
+ * | decorator  | \f$O(1)\f$  | \f$O(\log(k))\f$ | \f$O(\log(k))\f$        | \f$O(k)\f$              |
+ * | vector     | \f$O(1)\f$  | \f$O(1)\f$       | \f$O(n)\f$              | \f$O(n)\f$              |
+ *
+ * ### Size Overhead
+ *
+ * The seqan3::gap_decorator_anchor_set comes with additional costs in \f$O(k)\f$
+ * in order to store the gap information (in contrast to a vector over a gapped alphabet that needs \f$O(n)\f$
+ * additional space because we need copy over the complete sequence).
+ *
+ * ### Implementation details
+ *
+ * This decorator stores a std::set over tuples of `(pos, cumulative_size)` where every entry represents one
+ * contiguous stretch of gaps. `pos` is the (virtual) insert position in the underlying range and `cumulative_size`
+ * is the length of that contiguous stretch of gaps plus the length of all preceding elements.
+ * Resolving random access requires logarithmic access into the set and inserting or removing a gap symbol additionally
+ * entails updating all subsequent elements in the set to preserve correct cumulative sizes.
+ *
+ */
+template <std::ranges::RandomAccessRange inner_type>
+    requires std::ranges::SizedRange<inner_type>
+class gap_decorator_anchor_set
+{
+private:
+    //!\brief The iterator that moves over the seqan3::gap_decorator_anchor_set.
+    class gap_decorator_anchor_set_iterator
+    {
+    private:
+        //!\brief Pointer to the underlying container structure.
+        typename std::add_pointer_t<gap_decorator_anchor_set const> host{nullptr};
+        //!\brief Stores the virtual position index for the seqan3::gap_decorator_anchor_set.
+        typename gap_decorator_anchor_set::size_type pos{0u};
+        //!\brief Stores the physical position in the ungapped/underlying view.
+        int32_t ungapped_view_pos{0u};
```

Comment:
Why is this not of size_type?

Reply:
Because in case there are gaps at the beginning I want to set `ungapped_view_pos` to -1 on construction which saves me an if branch in the increment operator.

Reply:
Ok, but shouldn't it still be 64bit?